### PR TITLE
Rotate custom moon texture the right way

### LINKED ISSUE #11901: Moon texture is upside down
##### Minetest version
```
5.4.1
```

##### OS / Hardware

Operating system: Pop!_OS 21.10
CPU: i7 7700

GPU model: GTX 1060
OpenGL version: 4.6.0

##### Summary
Sun rising (about 6:00, ignore the size)
imagen

Moon rising (about 20:00, same texture)
imagen

##### Steps to reproduce
Apply the same texture both to sun and moon. Moon is upside down

### PR DESCRIPTION
**Pull request by Zughy**
Fixes #11901 
~~I've also scaled it down a bit, because when changed into a custom texture, the texture was almost twice as the original moon (halo of the original one included)~~ Never mind, scaling is correct, the PR only modifies the rotation

Before
imagen

After
imagen

## To do

This PR is Ready for Review.

## How to test
Put a custom texture on your moon

### CODE REVIEW

File: src/client/sky.cpp
Comment (appgurueu) [inline src/client/sky.cpp:678]: The scaling seems rather arbitrary to me?

**Comment by appgurueu** [comment]
I am mildly concerned about backwards-compatibility, but also not aware of any non-rotateable moon textures around.

File: src/client/sky.cpp
Comment (Zughy) [reply src/client/sky.cpp]: Having a better looking, there was another subtle halo, so yes, this particular modification doesn't make sense. Reverted, thanks

**Comment by Zughy** [comment]
Backward compatibility for a broken texture nobody ever reported? It seems rather trivial

**Review by SmallJoker — APPROVED** [review]
Fine by me.

**Comment by jordan4ibanez** [comment]
Team Rocket will love this change judging by that moon

**Comment by Desour** [comment]
Nice, more to do depending on protocol version or other indicators. Without even any documentation about this incompatibility. And texture packs will need to provide different textures depending on version.

What even is the "right" orientation for the textures?
One could think that the one used is at some point in time the same as for the sky cube map.

The real fix would be to add documentation.

Note that I'm a bit neutral because I haven't really used the ruben moon texture feature yet.

**Comment by appgurueu** [comment]
> The real fix would be to add documentation.

Indeed. Modders could do a simple `^[transformR180` if they want the moon texture to be rotated the way this PR does it.

**Comment by Zughy** [comment]
@Desour https://github.com/minetest/minetest/issues/11901#issuecomment-1002766165

**Comment by Desour** [comment]
> @Desour [#11901 (comment)](https://github.com/minetest/minetest/issues/11901#issuecomment-1002766165)

I have read the comments of the issue before of course.

As you are not giving any more information about what to do with that particular comment, I feel free to guess your intents: you want me to tell you what is wrong with your argumentation. Prepare for bikeshedding.

> If nobody has ever noticed that the moon was upside down (there are no issues in here), is either because they didn't notice - say, they put a symmetrical picture - or because they didn't care enough to open an issue. Either way, saying that it can't be fixed because it's already in a stable release, it's like saying "yes, 2 + 2 should be equal to 4, but it's too late, people can find a workaround to do sums". Also, we're not talking about something that will make game crashes.

You are arguing as if the moon texture orientation is a clear bug. But it's actually just a convention.
It's likely that some people did notice how the moon texture is oriented, but they didn't file an issue. They just found out by trial-and-error and rotated their textures according to the findings. These people probably didn't want to get their new stuff broken in the next release because the convention suddenly changes. Anyways, this are also just guesses.
And while changing this particular convention (probably) won't cause crashes, it still breaks things.

> To common sense. If the sun rises displaying the "R" in the right order, I expect the rising moon to display the same identical rotation

This is the case if you imagine that the sky cube was drawn on by drawing on a large stripe of paper and wrapping this paper around the 4 sides of the cube that are parallel to the rotation axis. But this is just one way of getting an orientation.
To common sense the simplest approach to get the orientation should be used, but your code changes that one function call to take other parameters than for the sun:
https://github.com/minetest/minetest/blob/75fc3a195a8a2a6b45338bff6fb72ec86c4398a0/src/client/sky.cpp#L631
Having the same rotation if the location is the same as the sun's would of course be a nice property, but I don't think it's worth breaking compatibility for this.

At the end, there's still nothing that stands against documentation.

**Comment by Zughy** [comment]
> At the end, there's still nothing that stands against documentation

`If you use the same texture for sun and moon, rotate the latter with '^[transformR180' to display it with the same angle`
"Why, engine limitation?"
"Oh no no, just because"

:\

**Comment by Wuzzy2** [comment]
Dumb question, but isn't the moon upside-down again when it's about to set? Or does it "flip" somehow after midnight?

I don't think this is really a bug that needs fixing. Just leave it as-is and document the behavior instead. No need to break compability for that. I'm already eagerly awaiting the players complaining the "smiley face moon" suddenly flipped after updating to MT 5.5.0. ;-)

To break compability, I'd say you need a stronger justification than "it's just weird".

Minetest REALLY needs to stop to constantly pull the rug under game makers and modders. I expect a game from 5.0.0 to work and look like before after the player upgraded to Minetest 5.5.0. If a game changes behavior or looks different after a Minetest upgrade, this is *bad*. It breaks game with no fault of the game maker.
Even if it sucks to say "you must rotate the moon by 180°" or whatever, preserving backwards-compability is much more important to me. Or just call the next version 6.0.0 instead of 5.5.0, then it's OK. :P
You might argue this particular thing is only be a small thing but you use this argument all the time. The small things add up.
Another example for when the rug was pulled under game makers was when `object_crosshair.png` was added with no fallback mechanism at all, so game makers were forced to touch their already *perfectly functioning and finished* game again to fix the crosshair.

**Comment by Zughy** [comment]
> Or does it "flip" somehow after midnight?

No it doesn't

> No need to break compability for that

The moon isn't the only problem. I've been playing with the whole celestial vault for about a week now, and every day I find something that doesn't work properly or that is poorly implemented: this, #11890 #11892 #11916 #11917 #11930. This tells me that the whole celestial vault hasn't been extensively tested by any contributor (whether is a modder, artist or core dev), because if they did, I wouldn't end up opening a new issue every day. We're talking about something that was implemented in [5.2.0](https://github.com/minetest/minetest/pull/8964#issuecomment-595397687), people had plenty of time to notice.  
On the contrary, if they did (and I'm not talking about the moon only), then it's on them they didn't point it out, as the project is open and anyone can speak their mind.  

In short: are we really breaking compatibility if no one ever noticed (or ignored it)? I don't think so

**Comment by appgurueu** [comment]
> In short: are we really breaking compatibility if no one ever noticed (or ignored it)? I don't think so. Also, worst case scenario, you enter an outdated server and you notice the non symmetrical moon is now upside down: I think everyone can live with that

It's a bold claim that "no one ever noticed". After more thought: No, I can't live with that, as this client-only change will make a 5.4.1 client see something different than a 5.5 one unless the modder works around this. Congrats, you have just replaced a simple `^[transformR180` workaround with more complex protocol checking and setting a different moon texture based on that, if the modder wants their non-symmetrical moon to show the same orientation across different clients.

**Comment by Zughy** [comment]
> It's a bold claim that "no one ever noticed"

Please don't cut sentences where it's convenient for you: I've faced bikeshedding already, I don't need strawmen too.  

Look, in all honesty I'm tired. I understand what you're saying, I'll let core devs decide, the PR is here, I did my part

**Comment by appgurueu** [comment]
> > It's a bold claim that "no one ever noticed"
> 
> Please don't cut sentences where it's convenient for you: I've faced bikeshedding already, I don't need strawmen too.

I have included the "or ignored it" part; it still remains a bold claim.

**Comment by Wuzzy2** [comment]
If it is decided that breaking compability is the way to go, *at least* put something into `minetest.features`, please.

**Comment by Zughy** [comment]
Apparently we're all forgetting that with or without my PR, the moon will eventually result rotated by 180° anyway. With my PR, this happens when falling, and without, when rising. I've just added coherence with the sun, I'm not breaking any compatibility.

Or, well, name me *one* game or mod (so texture packs are excluded) that has a texture of a non symmetrical moon on the Y axis featured inside, and that needs the moon with a certain rotation when rising

**Comment by sfan5** [comment]
> If it is decided that breaking compability is the way to go, _at least_ put something into `minetest.features`, please.

That wouldn't even help (except for singleplayer) because this change depends on the client features.

**Comment by erlehmann** [comment]
**I think that this patch breaks things and should be rejected.**

Rationale:

1. Every moon phase mod I have found uses textures that look wrong when rotated by 180 degrees.
2. If this patch was accepted, it would require every moon phase mod to un-rotate that texture again.
3. Nevertheless, even if mods do that, it seems impossible to me to show the same moon for older clients.

Therefore, the main thing that this patch does is introduce in incompatibility and create additional work.

I think this patch would break moon phases in the following mods:

* astral https://content.minetest.net/packages/cronvel/astral/
* moon_phases https://content.minetest.net/packages/TestificateMods/moon_phases/
* mcl_moon (part of MineClone2, MineClone5, Mineclonia)

I think this patch breaks moon phases in the following games & modpacks:

* Regional Weather
* Hamlet's Quest
* MineClone2
* MineClone5
* MineClonia

@Zughy @appgurueu @SmallJoker did you even test this with any mod that uses moon phases?

Additionally, the following games use a moon texture that looks noticeably different when rotated by 180 degrees:

* Exile
* Repixture

**Comment by Zughy** [comment]
Ok, do you want stupid bikeshedding (as apparently it's a thing in this topic) of your thesis, as you want to talk about realism? Let's do that.
Each one of those moon mods doesn't make sense in the first place. Yes, they currently rise in the correct way, but they don't when they go down. You know why? Because the moon doesn't do a 180° in real life. In fact, it does this: https://yewtu.be/watch?v=m708blgDucM .  This means that both with and without this PR, those mods are wrong *no matter what* for half of the night.  

Now, why exactly the rising part should be more important than the falling one? How am I supposed to understand in which phase I find myself into if I watch the sky in the middle of the night? Do I have to take into account the movement of the moon, so to calculate the right phase, remembering that the right one is the one rising? This sounds inconvenient to say the least.  

"Curious" to see if anyone implemented a system where the moon magically rotates its texture when reaching the highest point in the sky, I've tried some of those games. And I can tell you that:
- Hamlet's quest is been [more than a year](https://forum.minetest.net/viewtopic.php?f=15&t=18207&sid=5b4648914795486ee49035f5d4801201&start=200) since it was abandoned (so why even try)
- Mineclone2 doesn't
- I've refused to go further because it takes time, and what if Minetest implements such rotation someday in the future? It can't because there are people who implemented a DIY solution?

So the final question is: have *you* even tested those content in the first place *and* understood what this PR does before spitting bold formatted sentences? Because I don't think you have

---

If earlier I defined myself tired, now I'm simply plain annoyed. This behaviour is the reason why we can't have good things in here: people seeing compatibility issues where there are not, clutching at straws with unusual skybox conceptions, requiring changes that are useless, doing likes circlejerking, and refusing to understand the concept that (maybe if I write it for the third time it works) **the moon will eventually result rotated by 180° anyway**. Today I wake up and I find both the previous comment and this other nonsense https://github.com/minetest/minetest/pull/11843#issuecomment-1007350675. I could have coded more things, I could have worked more, I could have rested more, I could have done whatever, but no, I had to waste my time demonstrating once again how stupid this thing is and how retrocompatibility does have nothing to do with this because I now find the "Controversial" tag. @SmallJoker there is nothing controversial in this, just people who apparently have fun getting in the way without understanding what the PR is about. No wonder people stop contributing. I'm muting notifications, congratulations to each one of you.

**Comment by kilbith** [comment]
These mods rely on broken behavior and should be expected to adapt to correct behavior in the engine. No biggie.

> I'm muting notifications, congratulations to each one of you.

**Thank you**. To be fair, I do think you annoy more than one person (me). Stop wasting everyone's time.

**Comment by appgurueu** [comment]
> If earlier I defined myself tired, now I'm simply plain annoyed. This behaviour is the reason why we can't have good things in here: people seeing compatibility issues where there are not, clutching at straws with unusual skybox conceptions, requiring changes that are useless, doing likes circlejerking, and refusing to understand the concept that (maybe if I write it for the third time it works) **the moon will eventually result rotated by 180° anyway**. Today I wake up and I find both the previous comment and this other nonsense [#11843 (comment)](https://github.com/minetest/minetest/pull/11843#issuecomment-1007350675). I could have coded more things, I could have worked more, I could have rested more, I could have done whatever, but no, I had to waste my time demonstrating once again how stupid this thing is and how retrocompatibility does have nothing to do with this because I now find the "Controversial" tag. @SmallJoker there is nothing controversial in this, just people who apparently have fun getting in the way without understanding what the PR is about. No wonder people stop contributing. I'm muting notifications, congratulations to each one of you.

Let me sum up the pros and cons of this PR:

Pros: Modders don't have to add `^[transformR180` / rotate the texture manually to when setting the moon texture if they want moonrise to be consistent with sunrise.

Cons: Breaks backwards compatibility. Restoring consistency between 5.4 and 5.5 clients is significantly harder than appending a `^[transformR180`. Your argument is that moon orientation isn't really being relied on currently, which is clearly incorrect.

I feel you are biased as - other than projects like MCL, which have a long history - you don't really have to care about backwards compatibility and instead prefer starting with a "clean" API. Well, with semver that will have to wait until 6.0, unfortunately.

**Comment by erlehmann** [comment]
> you want to talk about realism?

Not at all. The mods & games I listed are not realistic.

They encode specific ideas about how a rising and setting moon looks like in the game.

If you look at the textures all those mods & games use, you can see that they are not symmetric at all.

This does, for example, make it possible to use the moon as a crude navigational aid.

Rotating the moon texture may make players wander in the opposite direction.

> the moon will eventually result rotated by 180° anyway

Yes and players die eventually anyway, so why not set their health to zero in the beginning of the game?

I agree with @Wuzzy here: The Minetest engine should be stable.

> I could have coded more things, I could have worked more, I could have rested more, I could have done whatever, but no, I had to waste my time

That is your own choice, @Zughy.

I have been in situations where a lot of devs did not understand my argument.

I may have thought they were stupid. But never once did I think that my argument was not controversial.

> These mods rely on broken behavior and should be expected to adapt to correct behavior

All mods that have moon textures without 180° rotational symmetry – i.e. literally all mods I found – rely on the current behaviour of the engine. Claiming that this is suddenly wrong when it worked without issues and it creates *a lot* more work to change it is a rhetorical trick.

> Stop wasting everyone's time.

I agree 100%. Just make the change opt-in and see if that gets you less opposition. Or rotate your textures, just like everyone else has been doing it for a while … instead of claiming everyone else was a total idiot.

**Comment by Anulo2** [comment]
From now on I will link this pr to everyone who will ask me "why should I not contribute to Minetest?". This is ridiculous. And I'm not talking about @Zughy statements.

**Comment by appgurueu** [comment]
> From now on I will link this pr to everyone who will ask me "why should I not contribute to Minetest?". This is ridiculous. And I'm not talking about @Zughy statements.

Your statement is ridiculous. This PR is not an improvement. It breaks compatibility solely based on which definition of moon orientation the PR author considers reasonable or "clean". This is not a contribution, it's pushing an opinionated decision.

**Comment by rubenwardy** [comment]
Everyone is getting a bit too heated in here

> From now on I will link this pr to everyone who will ask me "why should I not contribute to Minetest?". This is ridiculous. And I'm not talking about @Zughy statements.

This isn't a good take, and brigading is unproductive. This PR has been mostly civil to this point, people just disagree on this issue. 

The thing is that this isn't definitely a bug, which is why there's disagreements. The moon will be upside down half of the time anyway. The problem is a potential inconsistency in the API, but as celeron55 pointed out on IRC this may not be so: "currently when you see the sun and the moon at the same time, the texture is oriented the same way (on the opposite sides from you)". I'm undecided about what the _correct_ behaviour should be, but I don't think it's hugely important as it can be justified either way

If it matters which way the moon faces, then this is a breaking change (albeit on the small size). Minetest has made small breaking changes in the past when not documented by lua_api, but it needs to be justified by the outcome. [Moon Phases](https://content.minetest.net/packages/TestificateMods/moon_phases/) is one such thing that will be broken, the texture will render the opposite way to 5.4.

If it doesn't matter, then what's the point of this PR?

I've spent much longer than I would have liked thinking and writing about this PR

**Comment by Wuzzy2** [comment]
Because Zughy called me out for not testing this (oops, you got me!), I have now tested this.

The only thing this PR does is that the moon is now rotated normally when rising but is flipped when setting.
Before the PR, the moon is flipped when rising and normal when setting.
So, the PR does what I already thought it does, no surprise here.

So in other words, I can achieve the exact same effect when applying a `^[transformR180` to the moon texture in `set_moon`. And yes, I actually tested it, there is literally no difference. Therefore, no engine change is neccessary, as it can be trivially replicated in Lua. That alone should be grounds for rejection.

I'm very sorry, Zughy, but I still don't see the point of this. Alright, so one thing I get is that it is kinda annoying that the current behavior is the "opposite" of the sun (normal when rising, flipped when setting). This is true but the decision which way is which is *arbitrary*, we can argue endlessly about which way is "right" and I don't care. I don't even care if it is currently inconsistent with the sun. I am a modder and game developer and I do not care at all which is the "default" moon orientation, especially since a simple `transform` can fix it anyway. But some default orientation has to be chosen, and it is best to stick with it unless we have a strong reason to change it. Just a "nicer-looking API" is a reason, but it's not a really strong reason IMO, especially no other modder seems to ask for it.

And sorry, I have to disagree, this PR *will* break existing mods for apparently no convincing reason. That's just a fact. The question is only if we care enough about this. I do care. Especially since the change is 100% replicatable with Lua, i.e. it's a weak justification for breakage (even a minor one).

Also, I don't understand the 180° argument, that the moon rotates anyway. That's not really true, this is just due to perspective. It's not like Minetest actually actively flips the texture when the moon is path its zenith. That would be a bug. Minetest does NOT do that. When the moon moves past its zenith, it doesn't look strange or anything.

Finally, in case you accuse us of completely missing the point again: So what *is* the point? I don't understand. In our defense, the 1st post doesn't provide a rationale/justification. :-( So please enlighten us!

**Comment by Zughy** [comment]
What I see are wall of texts for something as simple as something that was explained in the very issue: sun rises with one inclination, moon with the other. I noticed that while testing my mod and I decided to fix that, because as a modder I expect the two celestial bodies to behave in the same way. It's a question of logic

> This does, for example, make it possible to use the moon as a crude navigational aid

No it doesn't, because it implies you should remember that the first quarter is going to be flipped past the zenith. Those mods are pretty useless when it comes to navigation (and I don't know about other uses, but I see no mods depending on them and this is out of the scope of the PR anyway).

> The "bugfix" label is just a disguise

Call it graphical improvement then, I personally don't care

> Moon Phases is one such thing that will be broken, the texture will render the opposite way to 5.4

For the fourth time: they are already flipped in the wrong direction for half of the night. The first quarter becomes the last quarter as soon as it passes the zenith.

Apparently you all believe that this thing is going to break the game experience somehow, so I deduce that the orientation when the moon rises *is* important to you. The only difference is that I noticed the incongruence with the sun and I fixed it. As there are no mods out there with a gameplay reason that needs the current rotation to remain untouched (because @erlehmann examples are either not gameplay related or just broken anyway, as I've showed twice), why all this big of a deal? And please do not turn this PR into "let's think about convoluted possible content where this thing might be a problem"

**Comment by erlehmann** [comment]
>> This does, for example, make it possible to use the moon as a crude navigational aid

> No it doesn't, because it implies you should remember that the first quarter is going to be flipped past the zenith.

Right now, players can deduce where north and south is based on a quick look at the moon in the astral mod.

You can test that: On the first night, the moon is much brighter on the south side than on the north side.

Rotating the texture, would, of course, mean that the moon would now point *north* instead of *south* here.

> why all this big of a deal

This thing is a big deal mainly because a lot of other people who develop on or maintain existing mods think that the Minetest engine should not break compatibility just to make a moon API call a bit more nicer for you.

From my point of view, your proposal looks super-weird: Even if there was no existing code to break, your patch removes the ability to look at the sun and moon textures being right-side-up at sunrise in a singlenode world – one of them will always be rotated the “wrong” way, i.e. be upside down. I hope that illustrates how arbitrary your and my own aesthetic judgments are.

**Comment by appgurueu** [comment]
> I noticed that while testing my mod and I decided to fix that, because as a modder I expect the two celestial bodies to behave in the same way. It's a question of logic

No. It's not a question of "logic". Both orientations can be regarded as "logical", as erlehmann has pointed out. It's purely a matter of opinion, and changing this opinion means breaking compat, which is unacceptable.

**Comment by Zughy** [comment]
> On the first night, the moon is much brighter on the south side than on the north side.

So, now, to bypass the "it's turned by 180° anyway" argument, we stopped referring to the world with left and right, using coordinates instead. And how exactly am I supposed to know that the moon is pointing north the first night? Or to complicate things, how am I supposed to not lose track of the amount of the nights passed to know where the current moon is pointing at? Again, navigation is inconvenient to say the least

> Both orientations can be regarded as "logical"

So if it's logical, why did you have to open a PR that explains in the docs how the moon texture works?

**Comment by appgurueu** [comment]
> > Both orientations can be regarded as "logical"
> 
> So if it's logical, why did you have to open a PR that explains in the docs how the moon texture works?

A right-handed coordinate system is as logical as a left-handed one. The docs currently don't mention that Minetest's coordinate system is left-handed. If I made a PR to document this, would that mean you could change it to be right-handed, because "it's more logical because I [you] didn't document it"?

We have the same situation with the moon orientation; I consider it more intuitive if sun and moon are rotated by 180° relative to each other at moon- or sunrise. Currently, the behavior isn't documented, but simply expected. I'm just documenting it; yours would need documentation too.

**Comment by Zughy** [comment]
> yours would need documentation too

How so? You all consider the rising moon more important than the falling one, thus I expect that one to be correctly displayed by default without rotating the picture, exactly as I do with the sun. That is what "logic" means.  Or we're back to the "imagine the celestial vault as a cylinder with stickers attached onto it"?

This thread looks like people that want to sound progressive, yet showing at the same time with their actions that they actually want to be conservative, finding whatever excuse they can. I'm so waiting for a core dev response, so we can at least move on

**Comment by Zughy** [comment]
Actually, as I fear this thing will keep going on forever and I really don't want that, can we please have a response, positive or negative as soon as any of you can (as it was debated on IRC yesterday anyway)?

@sfan5 @rubenwardy @celeron55 @v-rob

**Comment by appgurueu** [comment]
> > yours would need documentation too
> 
> How so? You all consider the rising moon more important than the falling one, thus I expect that one to be correctly displayed by default without rotating the picture, exactly as I do with the sun. That is what "logic" means.

That's what "logic" means to you.

> Right now is just arbitrary. Or we're back to the "imagine the celestial vault as a cylinder with stickers attached onto it"?

Yours is just as arbitrary. The celestial vault image you describe is not a step back, it's a solid model - other than your PR.

> This thread looks like people that want to sound progressive, yet showing at the same time with their actions that they actually want to be conservative, finding whatever excuse they can. I'm so waiting for a core dev response, so we can at least move on

If not breaking compatibility is "conservative" to you, then I am conservative.

**Comment by celeron55** [comment]
I think compatibility is most important here. The difference between the sun and the moon is arguably silly, but it's easy to document and work around if you need to. This is not something that will make or break Minetest for anyone.

There's a large chance the sky stuff will get some heavy rework at some point anyway at which point the orientations of things can be re-thought. I would hope it would get replaced by a more generic system so that we don't even need to call it a "celestial bodies system" but just "getting stuff drawn on the sky", but we'll see.

So, I would like to close this issue. No merge.

**Comment by sfan5** [comment]
Agreed from me.

**Pull request closed without merging**
